File: apps/cli/src/daemon-install.test.ts
import { describe, expect, test } from "bun:test";
import { createHash } from "node:crypto";

import type { FetchLike } from "./client";
import {
  type DaemonInstallFs,
  installLooperdBinary,
  resolveLooperdTarget,
} from "./daemon-install";

describe("daemon install helpers", () => {
  test("detects supported looperd targets from platform and arch", () => {
    expect(resolveLooperdTarget("darwin", "arm64")).toBe("darwin-arm64");
    expect(resolveLooperdTarget("darwin", "x64")).toBe("darwin-x64");
  });

  test("throws for unsupported platform and arch", () => {
    expect(() => resolveLooperdTarget("linux", "x64")).toThrow(
      "Unsupported platform/arch for looperd install: linux-x64. Supported targets: darwin-arm64, darwin-x64",
    );
  });

  test("fetches release metadata, selects exact asset and installs binary", async () => {
    const writes: Array<{ path: string; content: Uint8Array }> = [];
    const chmods: Array<{ path: string; mode: number }> = [];
    const renames: Array<{ from: string; to: string }> = [];
    const fetchCalls: string[] = [];
    const binaryBytes = new Uint8Array([1, 2, 3]);
    const checksum = createHash("sha256").update(binaryBytes).digest("hex");

    const fs: Partial<DaemonInstallFs> = {
      statImpl: async () => {
        throw new Error("missing");
      },
      mkdirImpl: async () => undefined,
      writeFileImpl: async (path, content) => {
        writes.push({ path: String(path), content: content as Uint8Array });
      },
      chmodImpl: async (path, mode) => {
        chmods.push({ path: String(path), mode: Number(mode) });
      },
      renameImpl: async (from, to) => {
        renames.push({ from: String(from), to: String(to) });
      },
    };

    const fetchImpl: FetchLike = async (input) => {
      const url = String(input);
      fetchCalls.push(url);

      if (url.includes("/releases/latest")) {
        return new Response(
          JSON.stringify({
            assets: [
              {
                name: "looperd-darwin-arm64",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64",
              },
              {
                name: "looperd-darwin-arm64.sha256",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64.sha256",
              },
            ],
          }),
        );
      }

      if (url === "https://example.invalid/looperd-darwin-arm64") {
        return new Response(binaryBytes);
      }

      if (url === "https://example.invalid/looperd-darwin-arm64.sha256") {
        return new Response(`${checksum}  looperd-darwin-arm64\n`);
      }

      throw new Error(`unexpected url ${url}`);
    };

    const result = await installLooperdBinary({
      fetchImpl,
      platform: "darwin",
      arch: "arm64",
      homeDir: "/Users/tester",
      force: false,
      fs,
    });

    expect(result).toEqual({
      target: "darwin-arm64",
      installPath: "/Users/tester/.looper/bin/looperd",
      downloadedFrom: "https://example.invalid/looperd-darwin-arm64",
      skipped: false,
    });
    expect(fetchCalls).toEqual([
      "https://api.github.com/repos/powerformer/looper/releases/latest",
      "https://example.invalid/looperd-darwin-arm64",
      "https://example.invalid/looperd-darwin-arm64.sha256",
    ]);
    expect(writes).toHaveLength(1);
    expect(writes[0]?.path).toBe("/Users/tester/.looper/bin/looperd.new");
    expect(Array.from(writes[0]?.content ?? [])).toEqual([1, 2, 3]);
    expect(chmods).toEqual([
      {
        path: "/Users/tester/.looper/bin/looperd.new",
        mode: 0o755,
      },
    ]);
    expect(renames).toEqual([
      {
        from: "/Users/tester/.looper/bin/looperd.new",
        to: "/Users/tester/.looper/bin/looperd",
      },
    ]);
  });

  test("is idempotent by default when looperd is already installed", async () => {
    const fetchImpl: FetchLike = async () => {
      throw new Error("fetch should not be called");
    };

    const result = await installLooperdBinary({
      fetchImpl,
      platform: "darwin",
      arch: "x64",
      homeDir: "/Users/tester",
      force: false,
      fs: {
        statImpl: async () => ({}) as never,
      },
    });

    expect(result).toEqual({
      target: "darwin-x64",
      installPath: "/Users/tester/.looper/bin/looperd",
      downloadedFrom: null,
      skipped: true,
    });
  });

  test("overwrites existing install when force is enabled", async () => {
    let downloadCount = 0;
    const binaryBytes = new Uint8Array([7]);
    const checksum = createHash("sha256").update(binaryBytes).digest("hex");

    const fetchImpl: FetchLike = async (input) => {
      const url = String(input);
      if (url.includes("/releases/latest")) {
        return new Response(
          JSON.stringify({
            assets: [
              {
                name: "looperd-darwin-x64",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-x64",
              },
              {
                name: "looperd-darwin-x64.sha256",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-x64.sha256",
              },
            ],
          }),
        );
      }

      if (url === "https://example.invalid/looperd-darwin-x64") {
        downloadCount += 1;
        return new Response(binaryBytes);
      }

      if (url === "https://example.invalid/looperd-darwin-x64.sha256") {
        return new Response(`${checksum}  looperd-darwin-x64\n`);
      }

      throw new Error(`unexpected url ${url}`);
    };

    const writes: string[] = [];
    await installLooperdBinary({
      fetchImpl,
      platform: "darwin",
      arch: "x64",
      homeDir: "/Users/tester",
      force: true,
      fs: {
        statImpl: async () => ({}) as never,
        mkdirImpl: async () => undefined,
        writeFileImpl: async (path) => {
          writes.push(String(path));
        },
        chmodImpl: async () => undefined,
        renameImpl: async () => undefined,
      },
    });

    expect(downloadCount).toBe(1);
    expect(writes).toEqual(["/Users/tester/.looper/bin/looperd.new"]);
  });

  test("reports clear error when metadata request fails", async () => {
    const fetchImpl: FetchLike = async () =>
      new Response("nope", {
        status: 503,
        statusText: "Service Unavailable",
      });

    await expect(() =>
      installLooperdBinary({
        fetchImpl,
        platform: "darwin",
        arch: "arm64",
        homeDir: "/Users/tester",
        force: false,
        fs: {
          statImpl: async () => {
            throw new Error("missing");
          },
        },
      }),
    ).toThrow(
      "Failed to fetch GitHub release metadata from https://api.github.com/repos/powerformer/looper/releases/latest (status 503 Service Unavailable)",
    );
  });

  test("reports clear error when binary asset download fails", async () => {
    const fetchImpl: FetchLike = async (input) => {
      const url = String(input);
      if (url.includes("/releases/latest")) {
        return new Response(
          JSON.stringify({
            assets: [
              {
                name: "looperd-darwin-arm64",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64",
              },
              {
                name: "looperd-darwin-arm64.sha256",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64.sha256",
              },
            ],
          }),
        );
      }

      return new Response("not found", {
        status: 404,
        statusText: "Not Found",
      });
    };

    await expect(() =>
      installLooperdBinary({
        fetchImpl,
        platform: "darwin",
        arch: "arm64",
        homeDir: "/Users/tester",
        force: false,
        fs: {
          statImpl: async () => {
            throw new Error("missing");
          },
        },
      }),
    ).toThrow(
      "Failed to download looperd binary from https://example.invalid/looperd-darwin-arm64 (status 404 Not Found)",
    );
  });

  test("rejects daemon install when checksum does not match downloaded binary", async () => {
    const fetchImpl: FetchLike = async (input) => {
      const url = String(input);
      if (url.includes("/releases/latest")) {
        return new Response(
          JSON.stringify({
            assets: [
              {
                name: "looperd-darwin-arm64",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64",
              },
              {
                name: "looperd-darwin-arm64.sha256",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64.sha256",
              },
            ],
          }),
        );
      }

      if (url.endsWith(".sha256")) {
        return new Response(`${"0".repeat(64)}  looperd-darwin-arm64\n`);
      }

      return new Response(new Uint8Array([1, 2, 3]));
    };

    await expect(() =>
      installLooperdBinary({
        fetchImpl,
        platform: "darwin",
        arch: "arm64",
        homeDir: "/Users/tester",
        force: true,
        fs: {
          statImpl: async () => ({}) as never,
        },
      }),
    ).toThrow("Downloaded looperd checksum mismatch");
  });

  test("cleans up temp file when atomic replace fails so retry is possible", async () => {
    const removed: string[] = [];
    const binaryBytes = new Uint8Array([9, 9, 9]);
    const checksum = createHash("sha256").update(binaryBytes).digest("hex");

    const fetchImpl: FetchLike = async (input) => {
      const url = String(input);
      if (url.includes("/releases/latest")) {
        return new Response(
          JSON.stringify({
            assets: [
              {
                name: "looperd-darwin-arm64",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64",
              },
              {
                name: "looperd-darwin-arm64.sha256",
                browser_download_url:
                  "https://example.invalid/looperd-darwin-arm64.sha256",
              },
            ],
          }),
        );
      }

      if (url.endsWith(".sha256")) {
        return new Response(`${checksum}  looperd-darwin-arm64\n`);
      }

      return new Response(binaryBytes);
    };

    await expect(() =>
      installLooperdBinary({
        fetchImpl,
        platform: "darwin",
        arch: "arm64",
        homeDir: "/Users/tester",
        force: true,
        fs: {
          statImpl: async () => ({}) as never,
          mkdirImpl: async () => undefined,
          writeFileImpl: async () => undefined,
          chmodImpl: async () => undefined,
          renameImpl: async () => {
            throw new Error("rename failed");
          },
          removeFileImpl: async (path) => {
            removed.push(String(path));
          },
        },
      }),
    ).toThrow("rename failed");

    expect(removed).toEqual(["/Users/tester/.looper/bin/looperd.new"]);
  });
});

File: internal/cliapp/daemon_install_test.go
package cliapp

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestResolveLooperdTarget(t *testing.T) {
	t.Parallel()

	target, err := resolveLooperdTarget("darwin", "arm64")
	if err != nil {
		t.Fatalf("resolveLooperdTarget(darwin, arm64) error = %v", err)
	}
	if target != "darwin-arm64" {
		t.Fatalf("resolveLooperdTarget(darwin, arm64) = %q, want %q", target, "darwin-arm64")
	}

	target, err = resolveLooperdTarget("linux", "amd64")
	if err != nil {
		t.Fatalf("resolveLooperdTarget(linux, amd64) error = %v", err)
	}
	if target != "linux-amd64" {
		t.Fatalf("resolveLooperdTarget(linux, amd64) = %q, want %q", target, "linux-amd64")
	}

	for _, arch := range []string{"amd64", "x64"} {
		_, err = resolveLooperdTarget("darwin", arch)
		if err == nil {
			t.Fatalf("resolveLooperdTarget(darwin, %s) error = nil, want unsupported", arch)
		}
	}
}

func TestInstallManagedDaemonInstallsBinary(t *testing.T) {
	t.Parallel()

	homeDir := t.TempDir()
	binary := []byte{1, 2, 3, 4}
	checksum := sha256.Sum256(binary)
	checksumText := hex.EncodeToString(checksum[:]) + "  looperd-darwin-arm64\n"

	app := New(Deps{
		HTTPClient: newTestHTTPClient(func(req *http.Request) (*http.Response, error) {
			switch req.URL.String() {
			case "https://api.github.com/repos/nexu-io/looper/releases/latest":
				return jsonResponse(t, http.StatusOK, `{"assets":[{"name":"looperd-darwin-arm64","browser_download_url":"https://example.invalid/looperd-darwin-arm64"},{"name":"looperd-darwin-arm64.sha256","browser_download_url":"https://example.invalid/looperd-darwin-arm64.sha256"}]}`), nil
			case "https://example.invalid/looperd-darwin-arm64":
				return binaryResponse(t, http.StatusOK, binary), nil
			case "https://example.invalid/looperd-darwin-arm64.sha256":
				return textResponse(t, http.StatusOK, checksumText), nil
			default:
				t.Fatalf("unexpected request URL %q", req.URL.String())
				return nil, nil
			}
		}),
		HomeDir:  homeDir,
		Platform: "darwin",
		Arch:     "arm64",
	})

	runtime := newCommandRuntime(app, nil)
	result, err := runtime.installManagedDaemon(context.Background(), false, "", nil)
	if err != nil {
		t.Fatalf("installManagedDaemon() error = %v", err)
	}

	installPath := filepath.Join(homeDir, ".looper", "bin", "looperd")
	if result.Target != "darwin-arm64" {
		t.Fatalf("result.Target = %q, want %q", result.Target, "darwin-arm64")
	}
	if result.InstallPath != installPath {
		t.Fatalf("result.InstallPath = %q, want %q", result.InstallPath, installPath)
	}
	if result.DownloadedFrom == nil || *result.DownloadedFrom != "https://example.invalid/looperd-darwin-arm64" {
		t.Fatalf("result.DownloadedFrom = %#v, want asset URL", result.DownloadedFrom)
	}
	if result.Skipped {
		t.Fatalf("result.Skipped = true, want false")
	}

	installedBytes, err := os.ReadFile(installPath)
	if err != nil {
		t.Fatalf("ReadFile(%q) error = %v", installPath, err)
	}
	if !bytes.Equal(installedBytes, binary) {
		t.Fatalf("installed bytes = %v, want %v", installedBytes, binary)
	}
	if _, err := os.Stat(installPath + ".new"); !os.IsNotExist(err) {
		t.Fatalf("temp install file exists unexpectedly: %v", err)
	}
}

func TestInstallManagedDaemonSkipsExistingBinary(t *testing.T) {
	t.Parallel()

	homeDir := t.TempDir()
	installPath := filepath.Join(homeDir, ".looper", "bin", "looperd")
	if err := os.MkdirAll(filepath.Dir(installPath), 0o755); err != nil {
		t.Fatalf("MkdirAll() error = %v", err)
	}
	if err := os.WriteFile(installPath, []byte("existing"), 0o755); err != nil {
		t.Fatalf("WriteFile() error = %v", err)
	}

	app := New(Deps{
		HTTPClient: newTestHTTPClient(func(req *http.Request) (*http.Response, error) {
			t.Fatalf("unexpected network request %q", req.URL.String())
			return nil, nil
		}),
		HomeDir:  homeDir,
		Platform: "darwin",
		Arch:     "arm64",
		RunCommand: func(ctx context.Context, command string, args []string, timeout time.Duration) (commandExecutionResult, error) {
			_ = ctx
			_ = timeout
			if command == installPath && strings.Join(args, " ") == "--version" {
				return commandExecutionResult{ExitCode: 0, Stdout: "1.2.3\n"}, nil
			}
			return commandExecutionResult{ExitCode: 1, Stderr: "not found"}, nil
		},
	})

	runtime := newCommandRuntime(app, nil)
	result, err := runtime.installManagedDaemon(context.Background(), false, "", nil)
	if err != nil {
		t.Fatalf("installManagedDaemon() error = %v", err)
	}
	if !result.Skipped {
		t.Fatalf("result.Skipped = false, want true")
	}
	if result.DownloadedFrom != nil {
		t.Fatalf("result.DownloadedFrom = %#v, want nil", result.DownloadedFrom)
	}
	if result.Target != "darwin-arm64" {
		t.Fatalf("result.Target = %q, want %q", result.Target, "darwin-arm64")
	}
}

func TestInstallManagedDaemonReportsNonExecutableExistingBinary(t *testing.T) {
	t.Parallel()

	homeDir := t.TempDir()
	installPath := filepath.Join(homeDir, ".looper", "bin", "looperd")
	if err := os.MkdirAll(filepath.Dir(installPath), 0o755); err != nil {
		t.Fatalf("MkdirAll() error = %v", err)
	}
	if err := os.WriteFile(installPath, []byte("existing"), 0o644); err != nil {
		t.Fatalf("WriteFile() error = %v", err)
	}

	app := New(Deps{HomeDir: homeDir, Platform: "darwin", Arch: "arm64"})
	runtime := newCommandRuntime(app, nil)
	_, err := runtime.installManagedDaemon(context.Background(), false, "", nil)
	if err == nil {
		t.Fatalf("installManagedDaemon() error = nil, want permission error")
	}
	for _, want := range []string{"not executable", "chmod +x " + installPath, "looper daemon install --force"} {
		if !strings.Contains(err.Error(), want) {
			t.Fatalf("error = %q, want to contain %q", err.Error(), want)
		}
	}
}

func TestInstallManagedDaemonReportsInvalidExistingBinary(t *testing.T) {
	t.Parallel()

	homeDir := t.TempDir()
	installPath := filepath.Join(homeDir, ".looper", "bin", "looperd")
	if err := os.MkdirAll(filepath.Dir(installPath), 0o755); err != nil {
		t.Fatalf("MkdirAll() error = %v", err)
	}
	if err := os.WriteFile(installPath, []byte("corrupt"), 0o755); err != nil {
		t.Fatalf("WriteFile() error = %v", err)
	}

	app := New(Deps{
		HomeDir:  homeDir,
		Platform: "darwin",
		Arch:     "arm64",
		RunCommand: func(ctx context.Context, command string, args []string, timeout time.Duration) (commandExecutionResult, error) {
			_ = ctx
			_ = timeout
			if command == installPath && strings.Join(args, " ") == "--version" {
				return commandExecutionResult{ExitCode: 1, Stderr: "exec format error"}, nil
			}
			return commandExecutionResult{ExitCode: 1, Stderr: "not found"}, nil
		},
	})
	runtime := newCommandRuntime(app, nil)
	_, err := runtime.installManagedDaemon(context.Background(), false, "", nil)
	if err == nil {
		t.Fatalf("installManagedDaemon() error = nil, want invalid binary error")
	}
	for _, want := range []string{"version check failed", "exec format error", "looper daemon install --force"} {
		if !strings.Contains(err.Error(), want) {
			t.Fatalf("error = %q, want to contain %q", err.Error(), want)
		}
	}
}

func TestDaemonInstallCommandPrintsHumanOutput(t *testing.T) {
	t.Parallel()

	stdout := &bytes.Buffer{}
	stderr := &bytes.Buffer{}
	binary := []byte{9, 8, 7}
	checksum := sha256.Sum256(binary)
	checksumText := hex.EncodeToString(checksum[:]) + "  looperd-darwin-arm64\n"

	app := New(Deps{
		Stdout: stdout,
		Stderr: stderr,
		HTTPClient: newTestHTTPClient(func(req *http.Request) (*http.Response, error) {
			switch req.URL.String() {
			case "https://api.github.com/repos/nexu-io/looper/releases/latest":
				return jsonResponse(t, http.StatusOK, `{"assets":[{"name":"looperd-darwin-arm64","browser_download_url":"https://example.invalid/looperd-darwin-arm64"},{"name":"looperd-darwin-arm64.sha256","browser_download_url":"https://example.invalid/looperd-darwin-arm64.sha256"}]}`), nil
			case "https://example.invalid/looperd-darwin-arm64":
				return binaryResponse(t, http.StatusOK, binary), nil
			case "https://example.invalid/looperd-darwin-arm64.sha256":
				return textResponse(t, http.StatusOK, checksumText), nil
			default:
				t.Fatalf("unexpected request URL %q", req.URL.String())
				return nil, nil
			}
		}),
		HomeDir:  t.TempDir(),
		Platform: "darwin",
		Arch:     "arm64",
	})

	exitCode := app.Run(context.Background(), []string{"daemon", "install"})
	if exitCode != 0 {
		t.Fatalf("Run([daemon install]) exit code = %d, want 0; stderr=%q", exitCode, stderr.String())
	}
	if !strings.Contains(stderr.String(), "Downloaded looperd (3 B)") {
		t.Fatalf("Run([daemon install]) stderr = %q, want download progress", stderr.String())
	}
	if !strings.Contains(stdout.String(), "Installed looperd (darwin-arm64) to ") {
		t.Fatalf("Run([daemon install]) stdout = %q, want install confirmation", stdout.String())
	}
	if !strings.Contains(stdout.String(), "Downloaded from https://example.invalid/looperd-darwin-arm64") {
		t.Fatalf("Run([daemon install]) stdout = %q, want download URL", stdout.String())
	}
}

func TestDownloadBinaryProgressFallsBackWhenLengthUnknown(t *testing.T) {
	t.Parallel()

	stderr := &bytes.Buffer{}
	app := New(Deps{
		HTTPClient: newTestHTTPClient(func(req *http.Request) (*http.Response, error) {
			if req.URL.String() != "https://example.invalid/looperd-darwin-arm64" {
				t.Fatalf("unexpected request URL %q", req.URL.String())
			}
			return &http.Response{
				StatusCode:    http.StatusOK,
				Status:        http.StatusText(http.StatusOK),
				Header:        http.Header{"Content-Type": []string{"application/octet-stream"}},
				Body:          io.NopCloser(strings.NewReader("abcd")),
				ContentLength: -1,
			}, nil
		}),
	})
	runtime := newCommandRuntime(app, nil)

	data, err := runtime.downloadBinary(context.Background(), "https://example.invalid/looperd-darwin-arm64", "looperd-darwin-arm64", stderr)
	if err != nil {
		t.Fatalf("downloadBinary() error = %v", err)
	}
	if string(data) != "abcd" {
		t.Fatalf("downloadBinary() data = %q, want abcd", string(data))
	}
	if !strings.Contains(stderr.String(), "Downloaded looperd (4 B)") {
		t.Fatalf("progress = %q, want unknown-size fallback", stderr.String())
	}
}

func TestDownloadBinaryProgressDoesNotReportSuccessOnReadError(t *testing.T) {
	t.Parallel()

	stderr := &bytes.Buffer{}
	app := New(Deps{
		HTTPClient: newTestHTTPClient(func(req *http.Request) (*http.Response, error) {
			if req.URL.String() != "https://example.invalid/looperd-darwin-arm64" {
				t.Fatalf("unexpected request URL %q", req.URL.String())
			}
			return &http.Response{
				StatusCode:    http.StatusOK,
				Status:        http.StatusText(http.StatusOK),
				Header:        http.Header{"Content-Type": []string{"application/octet-stream"}},
				Body:          errReadCloser{Reader: strings.NewReader("abcd"), err: io.ErrUnexpectedEOF},
				ContentLength: -1,
			}, nil
		}),
	})
	runtime := newCommandRuntime(app, nil)

	_, err := runtime.downloadBinary(context.Background(), "https://example.invalid/looperd-darwin-arm64", "looperd-darwin-arm64", stderr)
	if err == nil {
		t.Fatal("downloadBinary() error = nil, want read error")
	}
	if !strings.Contains(err.Error(), io.ErrUnexpectedEOF.Error()) {
		t.Fatalf("downloadBinary() error = %q, want %q", err.Error(), io.ErrUnexpectedEOF)
	}
	if strings.Contains(stderr.String(), "Downloaded looperd") {
		t.Fatalf("progress = %q, did not expect success line after read failure", stderr.String())
	}
	if !strings.Contains(stderr.String(), "Downloading looperd…") {
		t.Fatalf("progress = %q, want initial download line", stderr.String())
	}
}

func TestMPBDownloadProgressAbortsKnownSizeReadError(t *testing.T) {
	t.Parallel()

	progress := newMPBDownloadProgressFactory(io.Discard)
	tracker, ok := progress.newTracker("looperd-darwin-arm64", 4).(*mpbDownloadProgressTracker)
	if !ok {
		t.Fatal("newTracker() did not return *mpbDownloadProgressTracker")
	}
	reader := tracker.wrap(errReadCloser{Reader: strings.NewReader("abcd"), err: io.ErrUnexpectedEOF})
	_, err := io.ReadAll(reader)
	if err == nil {
		t.Fatal("ReadAll() error = nil, want read error")
	}
	if tracker.bar.Completed() {
		t.Fatal("tracker.bar.Completed() = true, did not expect completed after read error")
	}
	tracker.finish(false)
	progress.close()
	if tracker.bar.Completed() {
		t.Fatal("tracker.bar.Completed() = true, did not expect completed after read error")
	}
}

type errReadCloser struct {
	io.Reader
	err error
}

func (r errReadCloser) Read(p []byte) (int, error) {
	n, err := r.Reader.Read(p)
	if err == io.EOF && r.err != nil {
		return n, r.err
	}
	return n, err
}

func (errReadCloser) Close() error { return nil }

type roundTripFunc func(*http.Request) (*http.Response, error)

func (fn roundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return fn(req)
}

func newTestHTTPClient(fn roundTripFunc) *http.Client {
	return &http.Client{Transport: fn}
}

func jsonResponse(t *testing.T, status int, body string) *http.Response {
	t.Helper()
	return &http.Response{
		StatusCode: status,
		Status:     http.StatusText(status),
		Header:     http.Header{"Content-Type": []string{"application/json"}},
		Body:       io.NopCloser(strings.NewReader(body)),
	}
}

func textResponse(t *testing.T, status int, body string) *http.Response {
	t.Helper()
	return &http.Response{
		StatusCode: status,
		Status:     http.StatusText(status),
		Header:     http.Header{"Content-Type": []string{"text/plain"}},
		Body:       io.NopCloser(strings.NewReader(body)),
	}
}

func binaryResponse(t *testing.T, status int, body []byte) *http.Response {
	t.Helper()
	return &http.Response{
		StatusCode:    status,
		Status:        http.StatusText(status),
		Header:        http.Header{"Content-Type": []string{"application/octet-stream"}},
		Body:          io.NopCloser(bytes.NewReader(body)),
		ContentLength: int64(len(body)),
	}
}
